### PR TITLE
(feat): Make thread names identifiable.

### PR DESCRIPTION
**Pull request by azabbasi**
# Checklist
- [ ] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-java/blob/main/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- I submitted this PR against the correct branch: 
  - [ ] This pull-request is submitted against the `main` branch. 

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/fileupload/FileUploadTask.java [hunk 1]
```diff
@@ -119,7 +119,7 @@ public FileUploadTask(HttpsTransportManager httpsTransportManager)
     @Override
     public void run()
     {
-        Thread.currentThread().setName(THREAD_NAME);
+        Thread.currentThread().setName(THREAD_FIXED_NAME);
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/fileupload/FileUploadTask.java:122]: The file upload task will be removed from our upcoming v2 release, so you can just skip renaming this thread

Comment (azabbasi) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/fileupload/FileUploadTask.java:122]: Yeah, there are two file upload tasks, this one is related to getting the SAS uri, not the upload action. That one is deprecated and I didn't touch it.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientConfig.java [hunk 2]
```diff
@@ -470,6 +473,23 @@ public String getModuleId()
...
     }
 
+    public String getDeviceClientUniqueIdentifier()
+    {
+        // Use device Id if present, use module Id if no device Id is present, use a unique Identifier if neither was set.
+        String identifierPrefix = getDeviceId();
+        if (identifierPrefix == null || identifierPrefix.isEmpty())
+        {
+            identifierPrefix = getModelId();
+            if(identifierPrefix == null || identifierPrefix.isEmpty())
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientConfig.java:483]:
```suggestion
            if (identifierPrefix == null || identifierPrefix.isEmpty())
```

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubSendTask.java [hunk 3]
```diff
@@ -36,7 +36,8 @@ public IotHubSendTask(IotHubTransport transport)
 
     public void run()
     {
-        Thread.currentThread().setName(THREAD_NAME);
+        String threadName = this.transport.getDeviceClientUniqueIdentifier() + "-" + "Cxn" + transport.getTransportConnectionId() + "-" + THREAD_NAME;
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubSendTask.java:39]: What does "Cxn" stand for in these thread names?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java [hunk 4]
```diff
@@ -86,6 +86,11 @@
     // Keys are deviceIds. Helps with getting configs based on deviceIds
     private final Map<String, DeviceClientConfig> deviceClientConfigs = new ConcurrentHashMap<>();
 
+    // Is only set when it's not a multiplexing client.
+    private DeviceClientConfig deviceClientConfig = null;
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:90]: There's already a getter for the default config (the only config when not multiplexing)

```getDefaultConfig()```

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java [hunk 5]
```diff
@@ -1202,10 +1209,18 @@ private void openAsync() throws TransportException
         }
 
         this.reactor = createReactor();
+
+        String runnerUniqueIdentifier = this.isMultiplexing
+                ? "Multiplexed" + this.transportUniqueIdentifier
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:1214]: Should there be a dash between "Multiplexed" and this.transportUniqueIdentifier?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java [hunk 6]
```diff
@@ -1202,10 +1209,18 @@ private void openAsync() throws TransportException
...
+                : this.deviceClientConfig.getDeviceClientUniqueIdentifier();
+
+        String reactorRunnerPrefix = this.hostName + "-" + runnerUniqueIdentifier + "-" + "Cnx" + this.connectionId;
         ReactorRunner reactorRunner = new ReactorRunner(
             this.reactor,
             this.listener,
             this.connectionId,
+            reactorRunnerPrefix,
+            "OpenAsync",
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:1223]: This reactor thread owns the connection, not just opening the connection. I'd consider getting rid of this "OpenAsync" postfix here.

**Comment by timtay-microsoft** [comment]
Just a heads up for when you merge: it is "identifiable" not "identifyable"

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubSendTask.java (hunk 3 above)
Comment (azabbasi) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubSendTask.java:39]: Connection

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java (hunk 4 above)
Comment (azabbasi) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:90]: ah ! nice, will use that

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java (hunk 6 above)
Comment (azabbasi) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:1223]: I see, Will do.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/transport/amqp/AmqpsConnection.java [hunk 7]
```diff
@@ -65,6 +66,17 @@
 
     private final SSLContext sslContext;
 
+    // This field is used to uniquely identifying a connection.
```
Comment (timtay-microsoft) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/transport/amqp/AmqpsConnection.java:69]:
```suggestion
    // This field is used to uniquely identify a connection.
```

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/transport/amqp/AmqpsConnection.java [hunk 8]
```diff
@@ -202,7 +215,10 @@ public void openAmqpAsync()
 
         log.debug("Starting amqp reactor thread...");
         AmqpReactor amqpReactor = new AmqpReactor(this.reactor);
-        ReactorRunner reactorRunner = new ReactorRunner(amqpReactor);
+
+        String reactorRunnerPrefix = this.hostName + "-Cxn" + this.connectionId;
+
+        ReactorRunner reactorRunner = new ReactorRunner(amqpReactor, reactorRunnerPrefix, "AmqpOpenAsync");
```
Comment (timtay-microsoft) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/transport/amqp/AmqpsConnection.java:221]: "AmqpOpenAsync" as a thread suffix makes it sound like this thread is only devoted to opening the connection. The thread owns the connection completely, though, from open, to send, to close.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java [hunk 9]
```diff
@@ -1202,10 +1209,18 @@ private void openAsync() throws TransportException
         }
 
         this.reactor = createReactor();
+
+        String runnerUniqueIdentifier = this.isMultiplexing
+                ? "Multiplexed" + "-" + this.transportUniqueIdentifier
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:1214]:
```suggestion
                ? "Multiplexed-" + this.transportUniqueIdentifier
```

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java [hunk 10]
```diff
@@ -625,6 +632,20 @@ public void sendMessages()
...
+    }
+
+    String getDeviceClientUniqueIdentifier() {
+        // If it's not a multithreaded transport layer, we will use the device configuration to get the device unique identifier.
+        if (!this.isMultiplexing) {
+            return this.hostName + "-" + this.getDefaultConfig().getDeviceClientUniqueIdentifier();
+        }
+
+        // If this is a multithread transport layer, we will use its unique identifier.
+        return this.hostName + "-" + "Multiplexed" + this.transportUniqueIdentifier;
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:646]:
```suggestion
        return this.hostName + "-Multiplexed-" + this.transportUniqueIdentifier;
```

File: device/iot-device-client/src/test/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransportTest.java [hunk 11]
```diff
@@ -1054,7 +1049,7 @@ public void openConnectionWithAmqps() throws TransportException
             {
                 mockedConfig.getProtocol();
                 result = IotHubClientProtocol.AMQPS;
-                new AmqpsIotHubConnection(mockedConfig, false);
+                new AmqpsIotHubConnection(mockedConfig,anyString, false);
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/test/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransportTest.java:1052]:
```suggestion
                new AmqpsIotHubConnection(mockedConfig, anyString, false);
```

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/AmqpFileUploadNotificationReceive.java [hunk 12]
```diff
@@ -201,7 +201,7 @@ public synchronized FileUploadNotification receive(long timeoutMs) throws IOExce
 
             log.info("Receiving on file upload notification receiver for up to {} milliseconds", timeoutMs);
 
-            new ReactorRunner(amqpReceiveHandler, "AmqpFileUploadNotificationReceiver").run(timeoutMs);
+            new ReactorRunner(amqpReceiveHandler, this.hostName,"AmqpFileUploadNotificationReceiver").run(timeoutMs);
```
Comment (timtay-microsoft) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/AmqpFileUploadNotificationReceive.java:204]:
```suggestion
            new ReactorRunner(amqpReceiveHandler, this.hostName, "AmqpFileUploadNotificationReceiver").run(timeoutMs);
```

**Review by timtay-microsoft — APPROVED** [review]

**Pull request merged**